### ISSUE TITLE
Cannot externally specify default c_args for subproject in any way

### ISSUE THREAD

**Issue by LunarLambda** [mesonbuild/meson#11002]
I have a project that contains some libraries as subprojects, cross-compiled for ARM.

Right now, every project has to duplicate the same set of default `c_args`, as there's no way for the main project to pass them to the subprojects. Specifying `c_args` in the `subproject()`'s `default_options` doesn't work, specifying `-Dsubproject:c_args=...` errors, and the `project()`'s `default_options`, `add_project_arguments(...)` and `-Dc_args` don't propagate either. (And on that note, specifying `c_args` in `default_options` in a subproject's `project()` call doesn't work either, but `add_project_arguments` does)

I'm avoiding use of `add_global_arguments` since the project itself (could) be used as a subproject, and I'm not wholly convinced that

```meson
if not meson.is_subproject()
   add_global_arguments(...)
endif
```

is a good idea.

I'm also not keen on putting these options into the cross files, because they're not actually *necessary* for successful cross-compilation, they're moreso considered 'good practice' (the flags in question are `-mthumb -ffunction-sections -fdata-sections`), plus with cross files it is hard to do 'composition' of options rather than overriding.

Not only is this unnecessary duplication, it also means that users have no way to provide their own flags, or override this 'recommendation'.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? **Cross**
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.) **Arch Linux**
* what Python version are you using e.g. 3.8.0 **3.10.8**
* what `meson --version` **0.63.3**
* what `ninja --version` if it's a Ninja build **1.11.1**

**Comment by LunarLambda** [comment]
Per https://github.com/mesonbuild/meson/blob/master/docs/markdown/Release-notes-for-0.63.0.md#compiler-options-can-be-set-per-subproject it should be possible to override *all* compiler options (of which c_args is one)

I noticed that specfiying `c_std` works, but `c_args` doesn't.

**Comment by eli-schwartz** [comment]
@xclaesse Seems like commit 8867fb7999e35c6713c46af64becfd1a08146a6a accidentally excluded options defined as global, because those are tracked separately -- dunno why we don't just inherit those ones in all compilers.

**Comment by xclaesse** [comment]
Right, that's a known (I guess only by me) limitation of the per-subproject compiler options, because from an implementation POV they are different. We definitely should fix this, but it requires more refactoring of how we access <lang>_args and <lang>_link_args options. At the time I wrote the patch it was non trivial.

**Comment by Woynert** [comment]
The same issue exists with defines too (and fixing this issue would fix that as well).

Doing this:

```
meson setup build -Dmysubproject:c_args=-DTEST
```

Will return:

```
  User defined options
    mysubproject:c_args: -DTEST
```

`c_args` will be successfully set at the subproject level and you can print it:

```
# in subprojects/mysubproject/meson.build
message(get_option('c_args'))
...
```

This will print: `Message: ['-DTEST']` (success? not quite.)

__Despite__  `c_args` being present it will be ignored and won't have any actual effect on the build (tested on meson 1.10.1).
The reason is explained by @eyalitki  at https://github.com/mesonbuild/meson/issues/13260#issuecomment-2494839944. (`c_args` is only applied once and not per subproject).
He made a PR to fix it at #13944   but it seems there was some confusion and was wrongly closed thinking [this release](https://mesonbuild.com/Release-notes-for-1-8-0.html#per-project-subproject-options-rewrite) fixed it but that only works for other options, not for `c_args`, for example:

```
meson configure -Dmysubproject:optimization=3
# This works because other options are handled differently than c_args.
```

c_args will be set but have no effect.

Right now you have to use `add_global_arguments`, but this isn't ideal as OP mentioned. Another option is to patch each subproject's `meson_options.txt` and `meson.build` to include your new flags and enabled them by a new option. But then, these new options may not be needed for the subprojects to successfully build and also will result in a lot of duplication (once on main project, once per subproject).

Additionally, issue https://github.com/mesonbuild/meson/issues/13260 mentions "I would expect to see the options propagate to the subproject __and be applied to all build targets.__", but with the minimal example `c_args` did propagate (you can print it and see it) but on practice it isn't applied to the build target (has no effect).

I have made a minimal reproduction project [here](https://github.com/Woynert/meson-minimal-repro1).
